frontend/model/contracts/group.js

```diff
@@ -1,252 +1,279 @@
-          const threshold = Math.ceil(state.proposals[data.proposalHash].threshold * memberCount)
-          if (state.proposals[data.proposalHash].against.length > memberCount - threshold) {
-            Vue.delete(state.proposals, data.proposalHash)
-          }
+    },
+    // NOTE: there is no way to withdraw a vote on a proposal. User can however change their vote, including changing it to be indifferent.
+    'gi.contracts/group/proposalWithdraw': {
+      validate: objectOf({
+        proposalHash: string
+      }),
+      process (state, { data, meta }) {
+        const proposal = state.proposals[data.proposalHash]
+        if (!proposal) {
+          console.error(`proposalWithdraw: no proposal for ${data.proposalHash}!`, data)
+        } else if (proposal.meta.username !== meta.username) {
+          // TODO: properly handle these error conditions!
+          console.error(`proposalWithdraw: proposal ${data.proposalHash} belongs to ${proposal.meta.username} not ${meta.username}!`)
+        } else {
+          // TODO: make sure this is a synchronous function, and if not handle it appropriately
+          proposal.status = STATUS_WITHDRAWN
+          archiveProposal(state, data.proposalHash)
         }
       }
-    }
-  },
-  'GroupRecordInvitation': {
-    validate: objectOf({
-      username: string,
-      inviteHash: string,
-      sentDate: string
-    }),
-    vuexModuleConfig: {
-      mutation: (state, { data }) => { state.invitees.push(data.username) }
-    }
-  },
-  'GroupDeclineInvitation': {
-    validate: objectOf({
-      username: string,
-      inviteHash: string,
-      declinedDate: string
-    }),
-    vuexModuleConfig: {
-      mutation: (state, { data }) => {
-        const index = state.invitees.findIndex(username => username === data.username)
-        if (index > -1) { state.invitees.splice(index, 1) }
+    },
+    'gi.contracts/group/invite': {
+      validate: objectOf({
+        inviteSecret: string, // NOTE: simulate the OP_KEY_* stuff for now
+        numInvites: number
+      }),
+      process (state, { data }) {
+        Vue.set(state.invites, data.inviteSecret, {
+          generated: data.numInvites,
+          responses: {},
+          status: 'valid'
+        })
       }
-    }
-  },
-  'GroupAcceptInvitation': {
-    validate: objectOf({
-      username: string,
-      identityContractId: string,
-      inviteHash: string,
-      acceptanceDate: string
-    }),
-    vuexModuleConfig: {
-      mutation: (state, { data }) => {
-        const index = state.invitees.findIndex(username => username === data.username)
-        if (index > -1) {
-          state.invitees.splice(index, 1)
-          Vue.set(state.profiles, data.username, {
-            contractID: data.identityContractId,
-            groupProfile: {}
-          })
+    },
+    'gi.contracts/group/inviteDecline': {
+      validate: objectOf({
+        inviteSecret: string // NOTE: simulate the OP_KEY_* stuff for now
```

Comment:
is this the invite code? why calling it secret? `inviteCode` seems more intuitive.

Reply:
Because it will literally be a temporary secret that gives permission to outsiders to write messages of a restricted type to the contract.